labels.go

```diff
@@ -186,6 +190,45 @@ func PrevLabel(s string, n int) (i int, start bool) {
 	return 0, n > 1
 }
 
+// Compare compares domains according to the canonical ordering specified in RFC4034
+// returns an integer value similar to strcmp
+// (0 for equal values, -1 if s1 < s2, 1 if s1 > s2)
+func Compare(s1, s2 string) int {
+	s1b := []byte(s1)
+	s2b := []byte(s2)
+
+	doDDD(s1b)
+	doDDD(s2b)
+
+	s1lend := len(s1)
+	s2lend := len(s2)
+
+	for i := 0; ; i++ {
+		s1lstart, end1 := PrevLabel(s1, i)
+		s2lstart, end2 := PrevLabel(s2, i)
+
+		if end1 && end2 {
+			return 0
+		}
+
+		s1l := string(s1b[s1lstart:s1lend])
+		s2l := string(s2b[s2lstart:s2lend])
+
+		if !equal(s1l, s2l) {
+			return strings.Compare(strings.ToLower(s1l), strings.ToLower(s2l))
```

Comment:
ack, this works, but is expense, I _think_ we tried to not have allocations in the other compare code (including coredns). Is that possible here?

Reply:
It'll only do this with one pair of labels, though it could be completely alloc-free, yes. I'll see about implementing this.